File: BM92_最长无重复子数组/javascript_solutions/WindowSolution.ts
export function maxLength(arr: number[]): number {
    let maxLength = 0;

    const map = new Object();

    for (let left =0, right = 0; right < arr.length; right++) {
        const element = arr[right];
        
        if(map[element]){
            map[element]++
        }else{
            map[element] = 1
        }

        while(map[element] > 1) {
            const leftValue = arr[left];
            map[leftValue]--;
            left++;
        }

        const currentLength = right - left + 1;

        maxLength = maxLength > currentLength ?  maxLength : currentLength;
    }

    return maxLength;
}

File: BM92_最长无重复子数组/python_solutions/window_solution.py
from typing import List

class Solution:
    def maxLength(self, arr: List[int]) -> int:
        left = 0
        hash_map = {}
        max_len = 0

        for right in range(len(arr)):
            value = arr[right]

            if hash_map.get(value, 0):
                hash_map[value] += 1
            else:
                hash_map[value] = 1

            while hash_map[value] > 1:
                hash_map[arr[left]] -= 1
                left += 1

            max_len = max(max_len, right - left + 1)

        return max_len
